### PR TITLE
Replace oracle entry point by lab entry point

### PR DESCRIPTION
**Pull request by J535D165**
This PR replaces the `asreview oracle` command by `asreview lab`. The first still works, but the latter is preferred. This should avoid confusion in the future with the oracle, exploration, simulation mode in the frontend. 

The PR also sorts the entry points. The subcommands at the top are 'lab', 'simulate', and 'simulate-batch'.

### CODE REVIEW

**Comment by qubixes** [comment]
I agree with the changes in principle. I think it could be more elegant (I was thinking that oracle could be a subclass of the new lab entry point with the warning in there for example).

File: asreview/__main__.py [hunk 1]
```diff
@@ -47,16 +64,22 @@ def _output_available_entry_points(entry_points):
...
     # Try to load the entry point if available.
-    if len(sys.argv) > 1 and sys.argv[1] in entry_points:
+    if len(sys.argv) > 1 and (sys.argv[1] in entry_points or sys.argv[1] == "oracle"):
+
+        # first argument is the subcommand
+        subcommand = sys.argv[1]
+
+        # replace 'oracle' by 'lab'
+        if subcommand == "oracle":
+            print("Warning: subcommmand 'oracle' is replaced by 'lab'.")
```
Comment (qubixes) [inline asreview/__main__.py:78]: logging.warning ?

**Comment by J535D165** [comment]
Great feedback! Implemented all suggestions. Can you have a final check?

**Comment by GerbrichFerdinands** [comment]
Looks good to me. 

Should "oracle" also be replaced by "lab" in the tests?

https://github.com/asreview/asreview/blob/2046f4340b15b7984fe17de57ea44cc24f7fa905/tests/test_simulate.py#L201-L209

**Comment by qubixes** [comment]
@GerbrichFerdinands There is no more oracle mode testing... since more than a year? So, it can be removed altogether as an argument.

**Comment by GerbrichFerdinands** [comment]
And even though oracle still works, I think this should also be updated to "lab" in the Docker file?

https://github.com/asreview/asreview/blob/2046f4340b15b7984fe17de57ea44cc24f7fa905/docker/asreview-lab/Dockerfile#L3

**Review by GerbrichFerdinands — CHANGES_REQUESTED** [review]
Looks good to me! 
I am thinking, completely getting rid of 'batch' will break some of the scripts for HPC computing. Those scripts are also used for a couple of papers. Shouldn't we keep 'batch' just as we hold on to 'oracle' for now as well?

File: asreview/utils.py [hunk 2]
```diff
@@ -176,15 +176,31 @@ def is_iterable(i):
 
 
 def list_model_names(entry_name="asreview.models"):
-    return [entry.name
-            for entry in pkg_resources.iter_entry_points(entry_name)]
+    return [*get_entry_points(entry_name)]
```
Comment (qubixes) [inline asreview/utils.py:179]: Maybe list(x) is more obvious to read?

**Review by qubixes — CHANGES_REQUESTED** [review]

**Comment by qubixes** [comment]
@GerbrichFerdinands There is also an advantage in having a different name in that you can use the older version asreview-simulations along side it. The new version is different regardless, so it will break your/my HPC scripts anyway.

**Comment by J535D165** [comment]
@GerbrichFerdinands Thanks for the great feedback! I removed the oracle reference in the tests and updated Docker.

File: asreview/entry_points/gui.py [hunk 3]
```diff
@@ -10,17 +12,30 @@
...
-def _oracle_parser(prog="oracle", description=DESCRIPTION_ORACLE):
+# deprecated oracle class
+class OracleEntryPoint(LABEntryPoint):
+    description = "Graphical user interface for ASReview. (Deprecated)"
+
+    def _pre_execute(self):
+        logging.warning("Warning: subcommmand 'oracle' is replaced by 'lab'.")
+
+
+def _lab_parser(prog="lab", description=DESCRIPTION_ORACLE):
```
[the remark this answers is not in the document]
Comment (J535D165) [reply asreview/entry_points/gui.py]: Thanks

File: asreview/utils.py (hunk 2 above)
Comment (J535D165) [reply asreview/utils.py:179]: Hmm, I love this new syntax...

File: asreview/entry_points/gui.py [hunk 4]
```diff
@@ -10,17 +12,30 @@
...
 
         main(argv)
 
 
-def _oracle_parser(prog="oracle", description=DESCRIPTION_ORACLE):
+# deprecated oracle class
+class OracleEntryPoint(LABEntryPoint):
+    description = "Graphical user interface for ASReview. (Deprecated)"
+
+    def _pre_execute(self):
```
[the remark this answers is not in the document]
Comment (J535D165) [reply asreview/entry_points/gui.py:34]: Some devs are complaining when super isn't in the first line. I will replace this with your suggestion. 

You might be right on the 'formerly oracle' thing. Let's remove that one as well.

File: asreview/entry_points/gui.py [hunk 5]
```diff
@@ -1,26 +1,42 @@
...
 
 HOST_NAME = "localhost"
 PORT_NUMBER = 5000
 
 
-class GUIEntryPoint(BaseEntryPoint):
+class LABEntryPoint(BaseEntryPoint):
     description = "Graphical user interface for ASReview."
 
+    def _pre_execute(self):
```
Comment (qubixes) [inline asreview/entry_points/gui.py:18]: This can now be removed, right?

Comment (J535D165) [reply asreview/entry_points/gui.py:18]: Yes. Thanks. I was starving..

**Review by qubixes — APPROVED** [review]

File: asreview/entry_points/gui.py (hunk 5 above)
Comment (qubixes) [reply asreview/entry_points/gui.py:18]: Yes, me too!

Comment (J535D165) [reply asreview/entry_points/gui.py]: Resolved. Successfully merged branch 'pizza' into master.

**Comment by J535D165** [comment]
Thanks @qubixes and @GerbrichFerdinands for all the feedback!

**Comment by Rensvandeschoot** [comment]
Great work!!

Can we have an error message whenever someone tries to start ASReview with the oracle statement? It might take some time before existing users are familiar with the new way of starting the software...

**Comment by J535D165** [comment]
This change is backwards compatible. A deprecation warning is given as well. I'm not planning to remove this subcommand in the upcoming 6 months. The release of version 1 might be a good moment to remove it.

**Pull request merged**